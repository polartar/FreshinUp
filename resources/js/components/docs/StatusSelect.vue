<script>
import FBtnStatus from 'fresh-bus/components/ui/FBtnStatus'
export default {
  extends: FBtnStatus,
  props: {
    value: {
      type: Number,
      default: 1
    },
    options: {
      type: Array,
      required: true
    }
  },
  data () {
    return {
      items: this.options.map(item => {
        return {
          id: item.value,
          label: item.text,
          color: item.color
        }
      })
    }
  },
  computed: {
    activeItem () {
      for (let i in this.items) {
        if (this.items[i].id === this.value) return this.items[i]
      }
      return null
    }
  }
}
</script>
