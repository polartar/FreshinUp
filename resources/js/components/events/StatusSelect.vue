<script>
import FBtnStatus from 'fresh-bus/components/ui/FBtnStatus'
export default {
  extends: FBtnStatus,
  props: {
    value: {
      type: Number,
      default: 1
    },
    options: {
      type: Array,
      required: true
    }
  },
  computed: {
    items () {
      // Why don't we just use FBtnStatus' ability to change the "options" data via the props
      const color = ['grey', 'warning', 'warning', 'warning', 'warning', 'warning', 'success', 'error', 'grey']
      return this.options.map(item => {
        return {
          id: item.id,
          label: item.name,
          color: color[item.id - 1]
        }
      })
    }
  }
}
</script>
