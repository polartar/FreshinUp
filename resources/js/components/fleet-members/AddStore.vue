<template>
  <div>
    <div class="px-2 py-2">
      <filter-sorter
        :types="types"
        @input="runFilters"
      />
    </div>
    <div
      class="px-2"
      style="border-top: 1px solid gainsboro;"
    >
      <f-data-table
        :headers="headers"
        :items="stores"
        :multi-item-actions="multipleItemActions"
        item-key="uuid"
        hide-actions
        select-all
        v-bind="$attrs"
        v-on="$listeners"
      >
        <template v-slot:item-inner-name="{ item }">
          <div style="position: relative;">
            <v-tooltip top>
              <template v-slot:activator="{ on, attrs }">
                <a
                  class="primary--text"
                  href="#"
                  @click.prevent="viewItem"
                >
                  {{ get(item, 'name') }}
                </a>
              </template>
              View fleet member details (new tab)
            </v-tooltip>
          </div>
        </template>
        <template v-slot:item-inner-tags="{ item }">
          <div>
            <v-chip
              v-for="(tag, index) of get(item, 'tags', [])"
              :key="index"
              color="secondary"
              text-color="white"
            >
              {{ tag.name }}
            </v-chip>
          </div>
        </template>
        <template v-slot:item-inner-state_of_incorporation="{ item }">
          <div class="grey--text">
            {{ get(item, 'state_of_incorporation') }}
          </div>
        </template>
        <template v-slot:item-inner-manage="{ item }">
          <v-btn
            :depressed="manageButtonLabel(item) !== 'Assign'"
            :disabled="!isEligible(item)"
            :class="manageButtonClass(item)"
            @click="onManageClicked(item)"
          >
            {{ manageButtonLabel(item) }}
          </v-btn>
        </template>
      </f-data-table>
    </div>
  </div>
</template>
<script>
import get from 'lodash/get'
import FDataTable from '@freshinup/core-ui/src/components/FDataTable'
import AddStoreFilterSorter from './AddStoreFilterSorter'

export const HEADERS = [
  { text: 'Fleet member', value: 'name' },
  { text: 'State of incorporation', value: 'state_of_incorporation' },
  { text: 'Tags', value: 'tags' },
  { text: 'Manage', value: 'manage' }
]

export const MULTIPLE_ITEM_ACTIONS = []
export default {
  components: {
    FDataTable,
    FilterSorter: AddStoreFilterSorter
  },
  props: {
    headers: { type: Array, default: () => HEADERS },
    multipleItemActions: { type: Array, default: () => MULTIPLE_ITEM_ACTIONS },
    event: { type: Object, required: true, default: () => ({}) },
    stores: { type: Array, default: () => [] },
    types: { type: Array, default: () => [] }
  },
  data () {
    return {
      showFilters: false
    }
  },
  methods: {
    get,
    runFilters (payload) {
      this.$emit('run-filter', payload)
    },
    viewItem (item) {
      this.$emit('manage', 'view', item)
      this.$emit('manage-view', item)
    },
    // TODO: methods need more attention at some point later in the future
    hasBookedAnEvent (member) {
      const stores = get(member, 'event_stores', [])
      return stores.findIndex(e => e.event_uuid !== this.event.uuid) !== -1
    },
    isEligible (member) {
      return !this.hasBookedAnEvent(member) && !member['has_expired_licences_docs']
    },
    isAssignedToThisEvent (member) {
      const stores = get(member, 'event_stores', [])
      return stores.findIndex(e => e.event_uuid === this.event.uuid) !== -1
    },
    isDeclined (member) {
      const stores = get(member, 'event_stores', [])
      return stores.findIndex(e => e.event_uuid === this.event.uuid && e.declined) !== -1
    },
    manageButtonLabel (item) {
      if (this.isEligible(item)) {
        if (this.isDeclined(item)) { return 'Declined' }
        if (!this.isAssignedToThisEvent(item)) { return 'Assign' } else { return 'Assigned' }
      }
      if (this.hasBookedAnEvent(item)) { return 'Booked' }
      if (item['has_expired_licences_docs']) { return 'Expired' }
      return ''
    },
    onManageClicked (item) {
      if (this.manageButtonLabel(item) === 'Assign') {
        this.$emit('manage', 'assign', item)
        this.$emit('manage-assign', item)
      }
    },
    manageButtonClass (item) {
      const label = this.manageButtonLabel(item)
      return {
        'primary': label === 'Assign',
        'blue-grey lighten-5': ['Expired', 'Booked'].includes(label),
        'blue-grey lighten-3 white--text': ['Declined', 'Assigned'].includes(label)
      }
    }
  }
}
</script>
