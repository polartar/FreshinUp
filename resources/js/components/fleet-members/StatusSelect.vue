<script>
import FBtnStatus from 'fresh-bus/components/ui/FBtnStatus'
export default {
  extends: FBtnStatus,
  props: {
    value: {
      type: Number,
      default: 1
    },
    options: {
      type: Array,
      required: true,
      default: () => ([])
    }
  },
  computed: {
    items () {
      return this.options.map(option => ({
        label: option.name,
        id: option.id,
        color: 'warning'
      }))
    }
  }
}
</script>
