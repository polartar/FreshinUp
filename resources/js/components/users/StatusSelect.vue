<script>
// TODO: should be replace with @freshinup/core-ui/src/components/FBtnStatus on version >= 1.54.0
import FBtnStatus from 'fresh-bus/components/ui/FBtnStatus'

const LEAD = 1
const PENDING_INVITATION = 2
const PENDING_REVIEW = 3
const APPROVED = 4
const ON_HOLD = 5
const COLORS = {
  [LEAD]: 'grey',
  [PENDING_INVITATION]: 'warning',
  [PENDING_REVIEW]: 'warning',
  [APPROVED]: 'success',
  [ON_HOLD]: 'error'
}
export default {
  extends: FBtnStatus,
  props: {
    value: {
      type: Number,
      default: 1
    },
    options: {
      type: Array,
      required: true,
      default: () => ([])
    }
  },
  computed: {
    items () {
      return this.options.map(option => ({
        label: option.name,
        id: option.id,
        color: COLORS[option.id] ? COLORS[option.id] : 'grey'
      }))
    }
  }
}
</script>
