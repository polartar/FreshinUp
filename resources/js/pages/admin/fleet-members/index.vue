<template>
  <div>
    <v-flex
      d-flex
      align-center
      justify-space-between
      ma-2
    >
      <h2 class="white--text">
        {{ pageTitle }}
      </h2>
      <v-flex text-xs-right>
        <v-btn
          slot="activator"
          color="primary"
          dark
          @click="storeNew"
        >
          Add New Fleet Member
        </v-btn>
      </v-flex>
    </v-flex>

    <search-filter-sorter
      without-filter-label
      without-sort-by
      @run="runFilter"
    />

    <store-list
      v-if="!isLoading"
      :stores="stores"
      :statuses="statuses"
      :is-loading="isLoading || isLoadingList"
      :rows-per-page="pagination.rowsPerPage"
      :page="pagination.page"
      :total-items="pagination.totalItems"
      :sort-by="sorting.sortBy"
      :descending="sorting.descending"
      @paginate="onPaginate"
      @change-status="changeStatus"
      @manage-view="storeView"
      @manage-edit="storeEdit"
      @manage-delete="deleteStore"
      @manage-multiple-delete="deleteStore"
      @change_status_multiple="changeStatusMultiple"
    />
    <v-dialog
      v-model="deleteDialog"
      max-width="500"
    >
      <simple-confirm
        :class="{ 'deleting': deletablesProcessing }"
        :title="deleteDialogTitle"
        ok-label="Yes"
        cancel-label="No"
        @ok="onSubmitDelete"
        @cancel="onCancelDelete"
      >
        <div class="py-5 px-2">
          <template v-if="deletablesProcessing">
            <div class="text-xs-center">
              <p class="subheading">
                Processing, please wait...
              </p>
              <v-progress-circular
                :rotate="-90"
                :size="200"
                :width="15"
                :value="deletablesProgress"
                color="primary"
              >
                {{ deletablesStatus }}
              </v-progress-circular>
            </div>
          </template>
          <template v-else>
            <p class="subheading">
              <span v-if="deletables.length < 2">Store</span>
              <span v-else> Stores</span>
              : {{ deleteTemp | formatDeleteTitles }}
            </p>
          </template>
        </div>
      </simple-confirm>
    </v-dialog>
  </div>
</template>

<script>
import { mapGetters, mapActions, mapState } from 'vuex'
import { deletables } from 'fresh-bus/components/mixins/Deletables'
import StoreList from '~/components/stores/StoreList.vue'
import simpleConfirm from 'fresh-bus/components/SimpleConfirm.vue'
import get from 'lodash/get'
import SearchFilterSorter from 'fresh-bus/components/search/filter-sorter.vue'

export default {
  layout: 'admin',
  components: {
    StoreList,
    simpleConfirm,
    SearchFilterSorter
  },
  filters: {
    formatDeleteTitles (value) {
      return value.map(item => item.name).join(', ')
    }
  },
  mixins: [deletables],
  data () {
    return {
      pageTitle: 'Fleet Member List',
      deleteDialog: false,
      lastFilterParams: {
        sort: '-created_at'
      },
      deleteTemp: []
    }
  },
  computed: {
    isLoadingList () {
      return get(this.$store, 'state.stores.pending.items', true)
    },
    ...mapGetters('stores', {
      stores: 'items',
      pagination: 'pagination',
      sorting: 'sorting',
      sortBy: 'sortBy'
    }),
    ...mapGetters('page', ['isLoading']),
    ...mapState('stores', ['sortables']),
    ...mapGetters('storeStatuses', { 'statuses': 'items' }),
    deleteDialogTitle () {
      return this.deleteTemp.length < 2 ? 'Are you sure you want to delete this store?' : 'Are you sure you want to delete the following stores?'
    }
  },
  methods: {
    ...mapActions('page', {
      setPageLoading: 'setLoading'
    }),
    storeNew () {
      alert('coming soon')
    },
    storeView (store) {
      alert('coming soon')
    },
    storeEdit (store) {
      alert('coming soon')
    },
    deleteStore (stores) {
      if (!Array.isArray(stores)) {
        stores = [stores]
      }
      this.deleteTemp = stores
      this.deleteDialog = true
    },
    async onSubmitDelete () {
      this.deletablesProcessing = true
      this.deletablesProgress = 0
      this.deletablesStatus = ''
      let dispatcheables = []

      this.deleteTemp.forEach((store) => {
        dispatcheables.push(this.$store.dispatch('stores/deleteItem', { getItems: false, params: { id: store.uuid } }))
      })

      let chunks = this.chunk(dispatcheables, this.deleteTempParrallelRequest)
      let doneCount = 0

      for (let i in chunks) {
        await Promise.all(chunks[i])
        doneCount += chunks[i].length
        this.deleteTempStatus = doneCount + ' / ' + this.deleteTemp.length + ' Done'
        this.deleteTempProgress = doneCount / this.deleteTemp.length * 100
        await this.sleep(this.deletablesSleepTime)
      }

      this.filterStores(this.lastFilterParams)
      await this.sleep(500)
      this.deletablesProcessing = false
      this.deleteDialog = false
    },
    onCancelDelete () {
      this.deleteDialog = false
      this.deleteTemp = []
    },
    changeStatus (status, store) {
      this.$store.dispatch('stores/patchItem', { data: { status }, params: { id: store.uuid } }).then(() => {
        this.filterStores(this.lastFilterParams)
      })
    },
    changeStatusMultiple (status, stores) {
      stores.forEach((store) => {
        this.changeStatus(status, store)
      })
    },
    onPaginate (value) {
      this.$store.dispatch('stores/setPagination', value)
      this.$store.dispatch('stores/getItems')
    },
    runFilter(params) {
      this.filterStores({
        name: params.term,
        ...this.lastFilterParams
      })
    },
    filterStores (params) {
      this.lastFilterParams = params
      this.$store.dispatch('stores/setSort', params.sort)
      this.$store.dispatch('stores/setFilters', {
        ...this.$route.query,
        ...this.lastFilterParams
      })
      this.$store.dispatch('stores/getItems')
    }
  },
  beforeRouteEnterOrUpdate (vm, to, from, next) {
    vm.setPageLoading(true)

    Promise.all([
      vm.$store.dispatch('stores/setFilters', {
        ...vm.$route.query
      }),
      vm.$store.dispatch('storeStatuses/getItems')
    ])
      .then(() => {
        vm.$store.dispatch('page/setLoading', false)
        if (next) next()
      })
  }
}
</script>
