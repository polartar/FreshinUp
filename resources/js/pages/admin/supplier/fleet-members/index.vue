<script>
  import Page, { INCLUDE } from '~/pages/admin/fleet-members/index.vue'
  import { mapGetters } from 'vuex'
  export default {
    extends: Page,
    computed: {
      ...mapGetters(['currentUser']),
      ...mapGetters('suppliers/stores', {
        stores: 'items',
        pagination: 'pagination',
        sorting: 'sorting',
        sortBy: 'sortBy',
        storesLoading: 'itemsLoading'
      }),
      newLink () {
        return '/admin/supplier/fleet-members/new'
      },
    },
    async beforeRouteEnterOrUpdate (vm, to, from, next) {
      vm.setPageLoading(true)
      await vm.$store.dispatch('currentUser/getCurrentUser')
      Promise.all([
        vm.$store.dispatch('suppliers/stores/setFilters', {
          ...vm.$route.query,
          ...this.lastFilterParams
        }),
        vm.$store.dispatch('suppliers/stores/getItems', {
          params: {
            supplierId: vm.currentUser.uuid
          }
        }),
        vm.$store.dispatch('storeStatuses/getItems')
      ])
        .then()
        .catch((error) => console.error(error))
        .then(() => {
          vm.setPageLoading(false)
          if (next) next()
        })
    },
    methods: {
      onPaginate (value) {
        this.$store.dispatch('suppliers/stores/setPagination', value)
        this.$store.dispatch('suppliers/stores/getItems', {
          params: {
            supplierId: this.currentUser.uuid
          }
        })
      },
      storeViewOrEdit (store) {
        this.$router.push({ path: `/admin/supplier/fleet-members/${store.uuid}/edit` })
      },
      filterStores (params) {
        this.lastFilterParams = params
        this.$store.dispatch('suppliers/stores/setFilters', {
          ...this.$route.query,
          ...this.lastFilterParams
        })
        this.$store.dispatch('suppliers/stores/getItems', {
          params: {
            supplierId: this.currentUser.uuid,
            include: INCLUDE
          }
        })
      }
    }
  }
</script>
