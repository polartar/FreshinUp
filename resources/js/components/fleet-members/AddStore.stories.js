import { storiesOf } from '@storybook/vue'

import AddStore from './AddStore'

import { FIXTURE_STORES } from '../../../../tests/Javascript/__data__/stores'
import { FIXTURE_EVENT } from '../../../../tests/Javascript/__data__/event'
import { FIXTURE_STORE_TYPES } from '../../../../tests/Javascript/__data__/storeTypes'

export const Default = () => ({
  components: { AddStore },
  data () {
    return {
      event: FIXTURE_EVENT
    }
  },
  template: `
    <v-container>
      <add-store :event="event"/>
    </v-container>
  `
})

export const WithData = () => ({
  components: { AddStore },
  data () {
    return {
      stores: FIXTURE_STORES,
      types: FIXTURE_STORE_TYPES,
      event: FIXTURE_EVENT
    }
  },
  template: `
    <v-container>
      <add-store
        :stores="stores"
        :store-types="types"
        :event="event"/>
    </v-container>
  `
})

storiesOf('FoodFleet|components/event/AddStore', module)
  .add('Default', Default)
  .add('WithData', WithData)
