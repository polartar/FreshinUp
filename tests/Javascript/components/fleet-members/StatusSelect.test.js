import { shallowMount, mount } from '@vue/test-utils'
import Component from '~/components/fleet-members/StatusSelect.vue'
import * as Stories from '~/components/fleet-members/StatusSelect.stories'
import { FIXTURE_FLEET_MEMBERS_STATUSES } from 'tests/__data__/fleet-members'

describe('components/fleet-members/StatusSelect', () => {
  describe('Visuals', () => {
    test('ReadOnly', async () => {
      const wrapper = mount(Stories.ReadOnly())
      await wrapper.vm.$nextTick()
      expect(wrapper.element).toMatchSnapshot()
    })
    test('Selected', async () => {
      const wrapper = mount(Stories.Selected())
      await wrapper.vm.$nextTick()
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('Props & Computed', () => {
    test('value', () => {
      const wrapper = shallowMount(Component)
      expect(wrapper.vm.value).toEqual(1)

      wrapper.setProps({
        value: 2
      })
      expect(wrapper.vm.value).toEqual(2)
    })
    test('options', () => {
      const wrapper = shallowMount(Component)
      expect(wrapper.vm.options).toHaveLength(0)

      wrapper.setProps({
        options: FIXTURE_FLEET_MEMBERS_STATUSES
      })
      expect(wrapper.vm.options).toEqual(FIXTURE_FLEET_MEMBERS_STATUSES)
    })
    test('items', () => {
      const wrapper = shallowMount(Component)
      expect(wrapper.vm.items).toHaveLength(0)

      wrapper.setProps({
        options: FIXTURE_FLEET_MEMBERS_STATUSES
      })
      const expectations = FIXTURE_FLEET_MEMBERS_STATUSES.map(option => ({
        label: option.name,
        id: option.id,
        color: 'warning'
      }))
      expect(wrapper.vm.items).toEqual(expectations)
    })
  })
})
