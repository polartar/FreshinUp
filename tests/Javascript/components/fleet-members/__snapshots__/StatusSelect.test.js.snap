// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/fleet-members/StatusSelect Visuals ReadOnly 1`] = `
<div
  class="container"
>
  <button
    class="v-btn theme--dark warning warning-text--text"
    type="button"
  >
    <div
      class="v-btn__content"
    >
      
  Draft
 
      <i
        aria-hidden="true"
        class="v-icon v-icon--right material-icons theme--dark"
      >
        arrow_drop_down
      </i>
    </div>
  </button>
  <div
    class="v-menu"
  >
    <div
      class="v-menu__content theme--light "
      style="max-height: auto; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
    >
       
      <div
        class="v-list theme--light"
        role="list"
      >
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Draft
            </div>
          </a>
        </div>
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Pending
            </div>
          </a>
        </div>
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Revision
            </div>
          </a>
        </div>
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Rejected
            </div>
          </a>
        </div>
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Approved
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/fleet-members/StatusSelect Visuals Selected 1`] = `
<div
  class="container"
>
  <button
    class="v-btn theme--dark warning warning-text--text"
    type="button"
  >
    <div
      class="v-btn__content"
    >
      
  Draft
 
      <i
        aria-hidden="true"
        class="v-icon v-icon--right material-icons theme--dark"
      >
        arrow_drop_down
      </i>
    </div>
  </button>
  <div
    class="v-menu"
  >
    <div
      class="v-menu__content theme--light "
      style="max-height: auto; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
    >
       
      <div
        class="v-list theme--light"
        role="list"
      >
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Draft
            </div>
          </a>
        </div>
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Pending
            </div>
          </a>
        </div>
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Revision
            </div>
          </a>
        </div>
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Rejected
            </div>
          </a>
        </div>
        <div
          role="listitem"
        >
          <a
            class="v-list__tile v-list__tile--link theme--light"
          >
            <div
              class="v-list__tile__title"
            >
              Approved
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
